pkg/cloud/cloud.go

```diff
@@ -88,3 +100,29 @@ func New(region, kmsEndpoint string, qps, burst, retryTokenCapacity int) (AWSKMS
 	client := kms.NewFromConfig(cfg, kmsOptFns...)
 	return client, nil
 }
+
+func addConfusedDeputyHeaders(sourceArn string) func(*smithymiddleware.Stack) error {
+	return func(stack *smithymiddleware.Stack) error {
+		return stack.Build.Add(smithymiddleware.BuildMiddlewareFunc("KMSConfusedDeputyHeaders", func(
+			ctx context.Context, in smithymiddleware.BuildInput, next smithymiddleware.BuildHandler,
+		) (smithymiddleware.BuildOutput, smithymiddleware.Metadata, error) {
+			req, ok := in.Request.(*smithyhttp.Request)
+			if ok {
+				sourceAccount, err := getSourceAccount(sourceArn)
```

Comment:
Instead of parsing account in the middleware, can we parse it before calling `cfg.APIOptions = append(cfg.APIOptions, addConfusedDeputyHeaders(sourceArn))` and return error there?